### DISCUSSION TITLE
Download CSV Timestamps should be in local time, not UTC

### DISCUSSION THREAD

**Discussion by PengBG** [home-assistant/frontend#20610] [Other feature requests]
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.
- [ ] I have tried reproducing the issue in [safe mode](https://www.home-assistant.io/blog/2023/11/01/release-202311/#restarting-into-safe-mode) to rule out problems with unsupported custom resources.

### Describe the issue you are experiencing

Download data in CSV format gives offset in header compared to Energy UI. This happens to all ranges of date selection.
Example CSV data for period April 01 - April 21:
![image](https://github.com/home-assistant/frontend/assets/160076694/06fb5b33-3132-4efb-aa55-420a51b1553b)
Example Energy UI for the same period:
![image](https://github.com/home-assistant/frontend/assets/160076694/3e355e4e-a049-4c46-9a11-d33579e8f5d8)

### Describe the behavior you expected

I expected to see in CSV that colomn D will be marked as "2024-04-01T00:00:00.000Z" to represent 1st day of April
It is exported as "2024-03-31T22:00:00.000Z" data in this colomn D relates to Energy UI for april 1st
If range is day, why we have in header time stamp that starts at 22:00:00... time?
Same happens if you export for a single day, offset is there, data corresponds to Energy UI, but in CSV is labeled wrong

### Steps to reproduce the issue

1. Go to Energy dasboard
2. Select period day/month
3. Click 3 dots
4. Select "Download data"
5. Open downloaded CSV file
6. Check colomn headers in file and compare to Energy UI from where you downloaded file
...

### What version of Home Assistant Core has the issue?

2024.4.3

### In which browser are you experiencing the issue with?

Chrome, Edge

### Which operating system are you using to run this browser?

Win 10 64-bit

**Comment by karwosts** [comment]
I believe this is expected as the timestamps are in UTC (implied by the `Z` at the end).

So what is April 1 at midnight in your time zone is March 31 at 22:00 in UTC.

**Comment by PengBG** [comment]
well that implies that all who use HA need to know a lot about scripting and time tones etc. Peronally I didn't know tha about z at the end. What is expected behaviour here would be, what I see in Energy UI and how its declared there, to see same data in CSV file, now that is not a case

**Comment by karwosts** [comment]
I don't necessarily disagree with you, but it was mandated during review of this feature that all downloaded times should be printed in UTC, I guess that is a common practice. 

So this is working as intended.

**Comment by PengBG** [comment]
OK, lets put it this way, I started using HA about 3-4 months ago, and slowly building my setup, now I came accross energy panel and I can tell you that is totally unusable in the way that is presented in UI right now. But since it collects data from my sensors, I expected to be able to use that data in some external program for analysis. So I saw that has a download log option, and I liked that. After reading this what you wrote, personally I can't understand that decision or request that all times should be exported as UTC, since not all HA work in that time zone. HA clearly has support for local time since I can see logs and events in UI inside HA based on my local time. But for some strange reason exported data is in UTC time? So now I need to find someone to write me a plugin or script that will convet that UTC time to my local time and that I can finally have exported data matching what I already see in HA user interface, really strange decision and basically renders HA export data unusable for 50-60 % of users who will not go in that dirrection to write (or can't) that piece of code for conversion. 
For me personally I will probably revert to some other monitoring system for power usage other than HA because of this "common practice"

So HA can keep all data as it probably does in UTC, but if I as a user request to download some data, I expect it to match to already shown to me as user in UI of HA. Otherwise I can't trust even what I see in UI if there is a difference and I will revert to other solution, got my point?

**Comment by karwosts** [comment]
Moved to discussions w/ updated title as it's more appropriate there.

**Comment by Saxin** [comment]
@karwosts - thanks for implementing this, much appreciated. :) maybe small note on top of CSV to give unaware users a "strong" hint that this **Z** implies UTC.
(spent most of my evening trying to understand why exported data doesn't match to what is shown on the graph, common expectation from export to CSV button is WYSIWYG)

**Comment by moritzthecat** [comment]
from UX perspective this deviation between UI and CSV is causing quite some trouble for unaware people esp. with local time of UTC+, because it moves aggregated data to previous period (e.g. month) if not adjusted. So all european users potentially affected. 
- entity time stamp of period data is from beginning of period making time zones cross time boundaries if UTC time not respected 
- deviation between GUI and exported data is confusing

so proposal from above comments is understood that all affected users have to write a script to convert exported CSV data for post processing to local time zone?

suggestions:
- move aggregated entity time stamp from beginning of period to near end of period with no time zone conflict; at end of period then UTC- users have the pain ;-)

better 
- option to export in local time (HA likely has possibilities to implement?)
